### PR TITLE
C typesupport

### PR DESCRIPTION
**Pull request by dirk-thomas**
Currently trying to make Windows happy...

### CODE REVIEW

**Comment by tfoote** [comment]
@wjwwood says that you're not exporting symbols; that's why you don't get a .LIB with your .DLL.

**Comment by dirk-thomas** [comment]
The problem that it needs to be:
* in a single compilation unit in order to be able to compare the pointer instead of the string value
* visible in the the library in order to use the symbol in other code parts
* a compile-time constant in order to use it as in a static initializer list

I have not yet found a way to achieve all of this at the same time on Windows.

**Comment by gerkey** [comment]
As we discovered earlier today, in C you just can't use static const variables in static initializers.  E.g., this code doesn't compile with gcc on Linux:

~~~
static const int a = 0;
static const int b = a;
~~~

...but it does compile with g++.

**Comment by dirk-thomas** [comment]
This will be continued once ros2/rmw_implementation#9 has been completed.

**Comment by dirk-thomas** [comment]
I have rebased and updated this - ready for review:
* http://ci.ros2.org/job/ci_linux/734/
* http://ci.ros2.org/job/ci_osx/610/
* http://ci.ros2.org/job/ci_windows_opensplice/803/
* http://ci.ros2.org/job/ci_windows_connext_static/64/
* http://ci.ros2.org/job/ci_windows_connext_dynamic/52/

Locally I got it to compile and link the following example: https://github.com/ros2/examples/compare/master...c_msgs but since it is a weird mixture of C and C++ it won't run.

I think this should be merged even if we don't have a working talker / listener example in C with rcl(c) yet.

File: rosidl_typesupport_introspection_c/include/rosidl_typesupport_introspection_c/message_introspection.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "rosidl_generator_c/message_type_support.h"
+
+#include "rosidl_typesupport_introspection_c/visibility_control.h"
+
+typedef struct
+// ROSIDL_TYPESUPPORT_INTROSPECTION_C_PUBLIC
```
Comment (wjwwood) [inline rosidl_typesupport_introspection_c/include/rosidl_typesupport_introspection_c/message_introspection.h]: Should this be dropped?

File: rosidl_typesupport_introspection_c/include/rosidl_typesupport_introspection_c/message_introspection.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  uint32_t offset_;
+  const void * default_value_;
+  size_t (* size_function)(const void *);
+  const void * (*get_const_function)(const void *, size_t index);
+  void * (*get_function)(void *, size_t index);
+  bool (* resize_function)(void *, size_t size);
+} rosidl_typesupport_introspection_c__MessageMember;
+
+typedef struct
+// ROSIDL_TYPESUPPORT_INTROSPECTION_C_PUBLIC
```
Comment (wjwwood) [inline rosidl_typesupport_introspection_c/include/rosidl_typesupport_introspection_c/message_introspection.h]: Same here?

File: rosidl_typesupport_introspection_c/include/rosidl_typesupport_introspection_c/message_introspection.h (hunk 1 above)
Comment (dirk-thomas) [inline rosidl_typesupport_introspection_c/include/rosidl_typesupport_introspection_c/message_introspection.h]: Yeah, I removed them for now. We can put them in when we need them down the road.

**Comment by wjwwood** [comment]
lgtm

**Pull request merged**